client/commands/query/get.go

```diff
@@ -51,65 +52,85 @@ func Get(key []byte, prove bool) (data.Bytes, uint64, error) {
 // height as in Get.  Additionally, it will return a validated merkle
 // proof for the key-value pair if it exists, and all checks pass.
-func GetWithProof(key []byte) (data.Bytes, uint64, *iavl.KeyExistsProof, error) {
+func GetWithProof(key []byte) (data.Bytes, uint64,
+	*iavl.KeyExistsProof, *iavl.KeyNotExistsProof, error) {
+
 	node := commands.GetNode()
+	cert, err := commands.GetCertifier()
+	if err != nil {
+		return nil, 0, nil, nil, err
+	}
+	return getWithProof(key, node, cert)
+}
+
+func getWithProof(key []byte, node client.Client, cert certifiers.Certifier) (data.Bytes, uint64,
+	*iavl.KeyExistsProof, *iavl.KeyNotExistsProof, error) {
 
 	resp, err := node.ABCIQuery("/key", key, true)
 	if err != nil {
-		return nil, 0, nil, err
+		return nil, 0, nil, nil, err
 	}
-	ph := int(resp.Height)
 
 	// make sure the proof is the proper height
 	if !resp.Code.IsOK() {
-		return nil, 0, nil, errors.Errorf("Query error %d: %s", resp.Code, resp.Code.String())
+		return nil, 0, nil, nil, errors.Errorf("Query error %d: %s", resp.Code, resp.Code.String())
 	}
-	// TODO: Handle null proofs
-	if len(resp.Key) == 0 || len(resp.Value) == 0 || len(resp.Proof) == 0 {
-		return nil, 0, nil, lc.ErrNoData()
+	if len(resp.Key) == 0 || len(resp.Proof) == 0 {
+		return nil, 0, nil, nil, lc.ErrNoData()
 	}
-	if ph != 0 && ph != int(resp.Height) {
-		return nil, 0, nil, lc.ErrHeightMismatch(ph, int(resp.Height))
+	if resp.Height == 0 {
+		return nil, 0, nil, nil, errors.New("Height returned is zero")
 	}
 
-	check, err := GetCertifiedCheckpoint(ph)
+	check, err := getCertifiedCheckpoint(int(resp.Height), node, cert)
 	if err != nil {
-		return nil, 0, nil, err
+		return nil, 0, nil, nil, err
+	}
+
+	if len(resp.Value) > 0 {
+		// The key was found, construct a proof of existence.
+		proof := new(iavl.KeyExistsProof)
+		err = wire.ReadBinaryBytes(resp.Proof, &proof)
+		if err != nil {
+			return nil, 0, nil, nil, err
+		}
+
+		// Validate the proof against the certified header to ensure data integrity.
+		err = proof.Verify(resp.Key, resp.Value, check.Header.AppHash)
+		if err != nil {
+			return nil, 0, nil, nil, err
+		}
+
+		return data.Bytes(resp.Value), resp.Height, proof, nil, nil
 	}
 
-	proof := new(iavl.KeyExistsProof)
+	// The key wasn't found, construct a proof of non-existence.
+	proof := new(iavl.KeyNotExistsProof)
 	err = wire.ReadBinaryBytes(resp.Proof, &proof)
 	if err != nil {
-		return nil, 0, nil, err
+		return nil, 0, nil, nil, err
 	}
 
-	// validate the proof against the certified header to ensure data integrity
-	err = proof.Verify(resp.Key, resp.Value, check.Header.AppHash)
+	// Validate the proof against the certified header to ensure data integrity.
+	err = proof.Verify(resp.Key, check.Header.AppHash)
 	if err != nil {
-		return nil, 0, nil, err
+		return nil, 0, nil, proof, errors.Wrap(err, "Couldn't verify proof")
 	}
 
-	return data.Bytes(resp.Value), resp.Height, proof, nil
+	return nil, resp.Height, nil, proof, lc.ErrNoData()
```

Comment:
Yeah I debated this in my head - see https://github.com/tendermint/basecoin/pull/218#issuecomment-321785491

Essentially I didn't want to change the existing semantics of this set of functions, which is: 'get me this value and return an error if it doesn't exist.' - the proof is a bonus.

If this function were called `isKeyExists()` then it would be a different story - but here the function was not able to 'get' - hence the error.